### PR TITLE
[DPA-1313]: feat(deployment): add link contract deployment

### PR DESCRIPTION
**Pull request by graham-chainlink**
Added a new changeset in shared package which deploys the link token.

JIRA: https://smartcontract-it.atlassian.net/browse/DPA-1313

### CODE REVIEW

File: deployment/common/changeset/deploy_link_token.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	}
+	return deployment.ChangesetOutput{AddressBook: newAddresses}, nil
+}
+
+func deployLinkTokenContract(
+	lggr logger.Logger,
+	chain deployment.Chain,
+	ab deployment.AddressBook,
+) (*deployment.ContractDeploy[*link_token.LinkToken], error) {
+	linkToken, err := deployment.DeployContract[*link_token.LinkToken](lggr, chain, ab,
```
Comment (cgruber) [inline deployment/common/changeset/deploy_link_token.go:36]: This is not a criticism of you but I really hate that we have a function which returns a result, but also has a side-effect of updating a supplied address-book. :/ That's a sort of mis-matched use of function and side effect I'm not happy with. Need to rethink it.

You're doing it right, here, I think.

Comment (AnieeG) [reply deployment/common/changeset/deploy_link_token.go:36]: Is the link deployment needed for test environments? If not, can we not use already deployed link contracts in various chains?

Comment (graham-chainlink) [reply deployment/common/changeset/deploy_link_token.go:36]: > / That's a sort of mis-matched use of function and side effect I'm not happy with. Need to rethink it.

Good to know! I was kind of following the establish pattern here as seen in the code.

> Is the link deployment needed for test environments? If not, can we not use already deployed link contracts in various chains?

My understanding is that this changeset will be used for future new chains as well? And we are building a common bunch of useful changesets?

Comment (AnieeG) [reply deployment/common/changeset/deploy_link_token.go:36]: Are we sure that this is the contract we can officially use as link contract for new chains? I was under impression this is just a test equivalent.

Comment (AnieeG) [reply deployment/common/changeset/deploy_link_token.go:36]: If this is the intended one, please ignore me

Comment (cgruber) [reply deployment/common/changeset/deploy_link_token.go:36]: > Good to know! I was kind of following the establish pattern here as seen in the code.

Yeah - established patterns aren't all "Good" yet. We have been evolving this under deadline pressure, so we're just now coming back to re-evaluate emerging patterns. You're not doing bad here, it's more you're echoing something that I need to rethink.

File: deployment/common/changeset/deploy_link_token.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	"fmt"
+
+	"github.com/smartcontractkit/chainlink/deployment"
+	"github.com/smartcontractkit/chainlink/deployment/common/changeset/internal"
+)
+
+var _ deployment.ChangeSet[uint64] = DeployLinkToken
+
+// DeployLinkToken deploys a link token contract to the chain identified by the chainSelector.
+func DeployLinkToken(e deployment.Environment, chainSelector uint64) (deployment.ChangesetOutput, error) {
```
Comment (cgruber) [inline deployment/common/changeset/deploy_link_token.go:13]: This is fine for now, but I think we want to move this over to the chainlink/deployment once we've stabilized it. Maybe.  Ideally stuff like this is reusable by other products' changesets when they're deploying to a chain for the first time. They can't reuse this in their deployment logic if it's here. 

But for this PR, and getting your sea legs, this is fine. We can move it over in a subsequent change. Have to find a good pattern for code organization for shared things there too, so better to do it separately.

**Review by cgruber — DISMISSED** [review]

File: deployment/common/changeset/internal/link_token.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+
+	"github.com/smartcontractkit/chainlink/deployment"
+	"github.com/smartcontractkit/chainlink/deployment/common/types"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/shared/generated/link_token"
+)
+
+// DeployLinkTokenContract deploys a link token contract to the provided chain.
+func DeployLinkTokenContract(
```
Comment (connorwstein) [inline deployment/common/changeset/internal/link_token.go]: nit could probably just inline this actually in the changeset, I know we do this elsewhere but we're thinking its overkill and intend to inline the deploy.go functions inside the changesets.

**Review by connorwstein — DISMISSED** [review]

File: deployment/common/changeset/internal/link_token.go (hunk 3 above)
Comment (graham-chainlink) [reply deployment/common/changeset/internal/link_token.go]: ah nice, i had it right in my first iteration but then i followed the MCMS and moved the logic in to `internal`, let me bring the original back as i prefer that too!

File: deployment/common/changeset/deploy_link_token.go (hunk 1 above)
Comment (graham-chainlink) [reply deployment/common/changeset/deploy_link_token.go:36]: > If this is the intended one, please ignore me

@AnieeG thanks, let me double check with someone to be sure

**Review by cgruber — APPROVED** [review]

File: deployment/common/changeset/deploy_link_token.go (hunk 1 above)
Comment (graham-chainlink) [reply deployment/common/changeset/deploy_link_token.go:36]: @AnieeG i doubled checked with Rodrigo, yes this is the one to use.

**Review by AnieeG — APPROVED** [review]

**Pull request merged**